### PR TITLE
Add segmenter dictionary data to datagen

### PR DESCRIPTION
**Pull request by makotokato**
I would like to move segmenter dictionaries to testdata. These TOML files are converted from binary files although testdata doesn't include Khmer, Lao and Burmese.

Although dictionary segmenter key/structure of this fix is the following by locale, please suggest me about better key/structure.
- segmenter/dictionary@1
  - ja (Chinese and Japanese)
  - km
  - lo
  - my
  - th

### CODE REVIEW

File: experimental/segmenter/src/dictionary.rs [hunk 1]
```diff
@@ -197,20 +216,7 @@ mod tests {
...
-        ) {
-            Ok(s) => s,
-            Err(_) => panic!("invalid dictionary data"),
-        };
-
-        let data = UCharDictionaryBreakDataV1 {
-            trie_data: CJ_DICTIONARY.as_zerovec(),
-        };
-        let payload = DataPayload::<UCharDictionaryBreakDataV1Marker>::from_owned(data);
+        let payload = get_payload("ja").unwrap();
```
Comment (sffc) [inline experimental/segmenter/src/dictionary.rs:219]: This is a good question (how to represent the CJK dictionary as a language tag). Let me think about it more.

File: experimental/segmenter/src/line.rs [hunk 2]
```diff
@@ -124,11 +125,14 @@ impl LineBreakSegmenter {
...
             .take_payload()?;
 
-        // TODO: Use `provider` parameter after we support loading dictionary data from the
-        // production-ready providers.
-        let inv_provider = icu_provider::inv::InvariantDataProvider;
-        let dictionary_payload = inv_provider
-            .load_resource(&DataRequest::default())?
+        let locale: Locale = ("th")
+            .parse()
+            .map_err(|_| DataError::custom("cannot parse locale string"))?;
```
Comment (sffc) [inline experimental/segmenter/src/line.rs:130]: Suggestion, here and elsewhere: use the macro (which expands to a const function)

```suggestion
        let locale = locale!("th");
```

File: provider/datagen/src/transform/segmenter/mod.rs [hunk 3]
```diff
@@ -703,6 +704,105 @@ impl IterableResourceProvider<SentenceBreakDataV1Marker> for SegmenterRuleProvid
...
+            || req.options.get_langid() == langid!("lo")
+            || req.options.get_langid() == langid!("my")
+            || req.options.get_langid() == langid!("th")
+        {
+            let toml_data = self
+                .source
+                .segmenter()?
+                .read_and_parse_toml::<SegmenterDictionaryData>(Self::get_toml_filename(
+                    &req.options,
+                )?)?;
```
Comment (sffc) [inline provider/datagen/src/transform/segmenter/mod.rs:760]: Suggestion: make `get_toml_filename` return `Option<&'static str>`, and then do

```suggestion
                .read_and_parse_toml::<SegmenterDictionaryData>(Self::get_toml_filename(
                    &req.options,
                ).ok_or_else(|| DataErrorKind::MissingResourceOptions.with_req(UCharDictionaryBreakDataV1Marker::KEY, req))?)?;
```

and then you can remove the redundant `if req.options.get_langid() == ...` condition in this function.

Note that you should be returning `DataErrorKind::MissingResourceOptions`, not `DataError::custom`

File: provider/datagen/src/transform/segmenter/mod.rs [hunk 4]
```diff
@@ -703,6 +704,105 @@ impl IterableResourceProvider<SentenceBreakDataV1Marker> for SegmenterRuleProvid
...
+        r.push(ResourceOptions::from(locale));
+        let locale: icu_locid::Locale = ("my")
+            .parse()
+            .map_err(|_| DataError::custom("Cannot parse locale string"))?;
+        r.push(ResourceOptions::from(locale));
+        let locale: icu_locid::Locale = ("ja")
+            .parse()
+            .map_err(|_| DataError::custom("Cannot parse locale string"))?;
+        r.push(ResourceOptions::from(locale));
+        Ok(r)
```
Comment (sffc) [inline provider/datagen/src/transform/segmenter/mod.rs:802]: Suggestion: this whole function can be written as
```suggestion
        Ok(vec![locale!("th"), locale!("km"), locale!("lo"), locale!("my"), locale!("ja")])
```

**Review by sffc — COMMENTED** [review]
This looks good. Just a few small comments. I will think more about how to handle the multiple languages.

**Review by robertbastian — CHANGES_REQUESTED** [review]

File: experimental/segmenter/src/word.rs [hunk 5]
```diff
@@ -40,11 +41,14 @@ impl WordBreakSegmenter {
...
             .take_payload()?;
 
-        // TODO: Use `provider` parameter after we support loading dictionary data from the
-        // production-ready providers.
-        let inv_provider = icu_provider::inv::InvariantDataProvider;
-        let dictionary_payload = inv_provider
-            .load_resource(&DataRequest::default())?
+        let locale: Locale = ("th")
+            .parse()
+            .map_err(|_| DataError::custom("cannot parse locale string"))?;
```
Comment (robertbastian) [inline experimental/segmenter/src/word.rs:46]: macro

**Review by robertbastian — CHANGES_REQUESTED** [review]
Looks good, just some nits.

I'd also recommend naming this PR "Add segmenter dictionary data to datagen", it ends up in testdata as a result of that.

File: experimental/segmenter/src/dictionary.rs [hunk 6]
```diff
@@ -166,10 +166,29 @@ impl<'l> DictionarySegmenter<'l> {
...
+    use icu_locid::Locale;
     use zerovec::ZeroSlice;
 
+    fn get_payload(
+        locale_str: &str,
+    ) -> Result<DataPayload<UCharDictionaryBreakDataV1Marker>, DataError> {
+        let provider = icu_testdata::get_provider();
+        let locale: Locale = (locale_str)
+            .parse()
+            .map_err(|_| DataError::custom("Cannot parse locale string"))?;
```
[the remark this answers is not in the document]
Comment (makotokato) [reply experimental/segmenter/src/dictionary.rs]: Thanks, I use `get_payload(locale!("ja")`.

**Review by robertbastian — APPROVED** [review]

**Pull request merged**